File: PythonSandbox/src/leetcode/lc200_number_of_islands.py
'''
https://leetcode.com/problems/number-of-islands/
'''
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        visited = [[False for _ in grid[0]] for _ in grid]
        #print("visited: ", visited)
        count = 0
        for y in range(len(grid)):
            for x in range(len(grid[0])):
                if grid[y][x] == "1" and visited[y][x] == False:
                    self.floodFillIsland(grid, y, x, visited)
                    count += 1
        
        return count

    
    def floodFillIsland(self, grid, y, x, visited):
        if y < 0 or y > len(grid)-1 or x < 0 or x > len(grid[0])-1:
            return 

        if grid[y][x] == "0" or visited[y][x] == True:
            return
            
        visited[y][x] = True # fill current cell
        self.floodFillIsland(grid, y-1, x, visited)
        self.floodFillIsland(grid, y, x-1, visited)
        self.floodFillIsland(grid, y+1, x, visited)
        self.floodFillIsland(grid, y, x+1, visited)
        return 
        

File: JavaSandbox/src/leetcode/LC200_NumberOfIslands.java
package leetcode;
//https://leetcode.com/problems/number-of-islands/description/

import java.util.Arrays;

public class LC200_NumberOfIslands {
    public int numIslands(char[][] grid) {
        Boolean[][] visited = new Boolean[grid.length][grid[0].length];
        for(Boolean[] row: visited) {
            Arrays.fill(row, false);
        }

        int numIslands = 0;

        for(int r=0; r < grid.length; r++) {
            for(int c=0; c < grid[0].length; c++) {
                Boolean beenVisited = visited[r][c];
                if(grid[r][c]=='1' && !beenVisited) {
                    coverIsland(grid, r, c, visited);
                    visited[r][c] = true;
                    numIslands += 1;
                }
            }
        }

        return numIslands;
    }

    public void coverIsland(char[][] grid, int r, int c, Boolean[][] visited) {
        if(r < 0 || r > grid.length-1 || c < 0 || c > grid[0].length-1) {
            return;
        }

        if(visited[r][c] == true || grid[r][c] == '0') {
            return;
        }

        visited[r][c] = true;

        coverIsland(grid, r-1, c, visited);
        coverIsland(grid, r, c-1, visited);
        coverIsland(grid, r+1, c, visited);
        coverIsland(grid, r, c+1, visited);
    }
}
